README.md
# A Basic CakePHP + VueJS Single Page Application

[![Build Status](https://travis-ci.com/ishanvyas22/cakephpvue-spa.svg?branch=develop)](https://travis-ci.com/ishanvyas22/cakephpvue-spa)
[![Latest Stable Version](https://poser.pugx.org/ishanvyas22/cakephpvue-spa/v/stable)](https://packagist.org/packages/ishanvyas22/cakephpvue-spa)
[![Total Downloads](https://poser.pugx.org/ishanvyas22/cakephpvue-spa/downloads)](https://packagist.org/packages/ishanvyas22/cakephpvue-spa)
[![License](https://poser.pugx.org/ishanvyas22/cakephpvue-spa/license)](https://packagist.org/packages/ishanvyas22/cakephpvue-spa)

A basic CakePHP + VueJS single page application. Provides basic skeleton/boilerplate to quickly get started with CakePHP and Vue.js.

## ❤️ Support The Development

**Do you like this project? Support it by donating:**

<a href="https://www.buymeacoffee.com/ishanvyas" target="_blank">
    <img src="https://www.buymeacoffee.com/assets/img/custom_images/purple_img.png" alt="Buy Me A Coffee" style="height: 41px !important;width: 174px !important;box-shadow: 0px 3px 2px 0px rgba(190, 190, 190, 0.5) !important;-webkit-box-shadow: 0px 3px 2px 0px rgba(190, 190, 190, 0.5) !important;" >
</a>

<a href="https://www.patreon.com/ishanvyas">
    <img src="https://c5.patreon.com/external/logo/become_a_patron_button@2x.png" width="160">
</a>

**or** [Paypal me](https://paypal.me/IshanVyas?locale.x=en_GB)

**or** [![Contact me on Codementor](https://www.codementor.io/m-badges/isvyas/get-help.svg)](https://www.codementor.io/@isvyas?refer=badge)

### Follow me

-   [GitHub](https://github.com/ishanvyas22)
-   [Instagram](https://www.instagram.com/ishancodes)
-   [LinkedIn](https://www.linkedin.com/in/ishan-vyas-314111112)
-   [Twitter](https://twitter.com/ishanvyas22)

## Installation

1. Get project into your system

    Via [composer](https://getcomposer.org/):

    ```bash
    composer create-project ishanvyas22/cakephpvue-spa
    ```

    Via cloning the project into your server:

    ```bash
    git clone git@github.com:ishanvyas22/cakephpvue-spa.git
    ```

2. Install composer dependencies
    ```bash
    composer install
    ```
3. Connect with your [database](https://book.cakephp.org/3.0/en/orm/database-basics.html#configuration)
4. Run the [migrations](https://book.cakephp.org/3.0/en/migrations.html#overview)
    ```bash
    bin/cake migrations migrate
    ```
5. Start [CakePHP web server](https://book.cakephp.org/3.0/en/installation.html#development-server) or [create v-host](https://www.digitalocean.com/community/tutorials/how-to-install-the-apache-web-server-on-ubuntu-18-04) to run this app.

## Development

1. Install npm packages
    ```bash
    npm install
    ```
2. Run below command (Uses awesome [Asset Mix](https://github.com/ishanvyas22/asset-mix) plugin for asset compilation)
    - **To watch every changes in your vue file**
        ```bash
        npm run watch
        ```
    - **To generate development version of js file**
        ```bash
        npm run dev
        ```
    - **To generate production version of js file**
        ```bash
        npm run prod
        ```

## Notes

-   This is project is under development currently so you might not want to use this app in production directly.
-   This is just a skeleton/example how you can implement the Vue.js with CakePHP.

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information about recent changes.

## Contribution

I love to welcome your contributions if you know CakePHP / Vue.js.

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Release Notes

## [Unreleased](https://github.com/ishanvyas22/cakephpvue-spa/compare/2.1.1...master)

## [2.1.1 (2021-11-14)](https://github.com/ishanvyas22/cakephpvue-spa/compare/2.1.0...2.1.1)

### Updated

-   Bump y18n from 4.0.0 to 4.0.1 ([#23](https://github.com/ishanvyas22/cakephpvue-spa/pull/23))
-   Bump ssri from 6.0.1 to 6.0.2 ([#24](https://github.com/ishanvyas22/cakephpvue-spa/pull/24))
-   Bump lodash from 4.17.19 to 4.17.21 ([#26](https://github.com/ishanvyas22/cakephpvue-spa/pull/26))
-   Bump composer/composer from 1.10.10 to 1.10.22 ([#25](https://github.com/ishanvyas22/cakephpvue-spa/pull/25))
-   Bump browserslist from 4.14.0 to 4.16.6 ([#27](https://github.com/ishanvyas22/cakephpvue-spa/pull/27))
-   Bump dns-packet from 1.3.1 to 1.3.4 ([#28](https://github.com/ishanvyas22/cakephpvue-spa/pull/28))
-   Bump ws from 6.2.1 to 6.2.2 ([#29](https://github.com/ishanvyas22/cakephpvue-spa/pull/29))
-   Bump path-parse from 1.0.6 to 1.0.7 ([#30](https://github.com/ishanvyas22/cakephpvue-spa/pull/30))
-   Bump color-string from 1.5.3 to 1.6.0 ([#32](https://github.com/ishanvyas22/cakephpvue-spa/pull/32))
-   Bump url-parse from 1.5.1 to 1.5.3 ([#31](https://github.com/ishanvyas22/cakephpvue-spa/pull/31))
-   Bump axios from 0.21.1 to 0.21.2 ([#33](https://github.com/ishanvyas22/cakephpvue-spa/pull/33))
-   Bump composer/composer from 1.10.22 to 1.10.23 ([#36](https://github.com/ishanvyas22/cakephpvue-spa/pull/36))

## [2.1.0 (2021-03-28)](https://github.com/ishanvyas22/cakephpvue-spa/compare/2.0.1...2.1.0)

### Fixed

-   npm security vulnerabilities ([a23625f](https://github.com/ishanvyas22/cakephpvue-spa/commit/a23625f45fb1c56598097a340ce30b99b1421da9))

### Removed

-   `post-autoload-dump` hook from `composer.json` ([ab2ac60](https://github.com/ishanvyas22/cakephpvue-spa/commit/ab2ac609c66972f9d1f57f45cf62ce416f0d72af))

### Updated

-   Bump `cakephp/plugin-installer` package to 1.3.1 ([e0cb837](https://github.com/ishanvyas22/cakephpvue-spa/commit/e0cb8378751757d74c8a1fb0047bfd38375ef5f0))
-   Bump ini from 1.3.5 to 1.3.8 ([#20](https://github.com/ishanvyas22/cakephpvue-spa/pull/20))
-   Bump axios from 0.19.2 to 0.21.1 ([#21](https://github.com/ishanvyas22/cakephpvue-spa/pull/21))
-   Bump elliptic from 6.5.3 to 6.5.4 ([#22](https://github.com/ishanvyas22/cakephpvue-spa/pull/22))

## [2.0.1 (2020-08-09)](https://github.com/ishanvyas22/cakephpvue-spa/compare/2.0.0...2.0.1)

### Changed

-   Changed package link to getcomposer.org ([c514d07](https://github.com/ishanvyas22/cakephpvue-spa/commit/c514d073889fc0273c0d59256d726d3fab2b1980))
-   Change build status badge position to at the top in `README.md` file ([bb334c5](https://github.com/ishanvyas22/cakephpvue-spa/commit/bb334c5e2dcf938dd9de8b99bf480ecd3d5a648c))
-   Bump cakephp/cakephp from 4.1.1 to 4.1.2 ([c533f0f](https://github.com/ishanvyas22/cakephpvue-spa/commit/c533f0fb753647475a8698cf3b0a27ed1dbdeb00))

### Removed

-   Unused html files ([c336172](https://github.com/ishanvyas22/cakephpvue-spa/commit/c336172a49a9130d2b393fbf50d056c25086408e))

## 2.0.0 (2020-08-08)

### Added

-   Added support for CakePHP 4.x ([#18](https://github.com/ishanvyas22/cakephpvue-spa/pull/18))
-   Added phpstan as a dev dependency by default ([1dff97e](https://github.com/ishanvyas22/cakephpvue-spa/pull/18/commits/1dff97e0159f0cba2d3ba81b7b1fcc3f3757c8c9))

### Changed

-   Update laravel-mix along with all package dependencies ([4a7dfac](https://github.com/ishanvyas22/cakephpvue-spa/pull/18/commits/4a7dfaca66eb8962f977c12b9a949d9a8726b008))

### Fixed

-   Rename CHANGLOG.md to CHANGELOG.md ([2eb3eae](https://github.com/ishanvyas22/cakephpvue-spa/pull/18/commits/2eb3eae7292cc8a083ba90245dff7ffbaca4cca1))

Also this release contains minor enhancements & bug fixes.


config/app.default.php
<?php

return [
    /**
     * Debug Level:
     *
     * Production Mode:
     * false: No error messages, errors, or warnings shown.
     *
     * Development Mode:
     * true: Errors and warnings shown.
     */
    'debug' => filter_var(env('DEBUG', true), FILTER_VALIDATE_BOOLEAN),

    /**
     * Configure basic information about the application.
     *
     * - namespace - The namespace to find app classes under.
     * - defaultLocale - The default locale for translation, formatting currencies and numbers, date and time.
     * - encoding - The encoding used for HTML + database connections.
     * - base - The base directory the app resides in. If false this
     *   will be auto detected.
     * - dir - Name of app directory.
     * - webroot - The webroot directory.
     * - wwwRoot - The file path to webroot.
     * - baseUrl - To configure CakePHP to *not* use mod_rewrite and to
     *   use CakePHP pretty URLs, remove these .htaccess
     *   files:
     *      /.htaccess
     *      /webroot/.htaccess
     *   And uncomment the baseUrl key below.
     * - fullBaseUrl - A base URL to use for absolute links. When set to false (default)
     *   CakePHP generates required value based on `HTTP_HOST` environment variable.
     *   However, you can define it manually to optimize performance or if you
     *   are concerned about people manipulating the `Host` header.
     * - imageBaseUrl - Web path to the public images directory under webroot.
     * - cssBaseUrl - Web path to the public css directory under webroot.
     * - jsBaseUrl - Web path to the public js directory under webroot.
     * - paths - Configure paths for non class based resources. Supports the
     *   `plugins`, `templates`, `locales` subkeys, which allow the definition of
     *   paths for plugins, view templates and locale files respectively.
     */
    'App' => [
        'namespace' => 'App',
        'encoding' => env('APP_ENCODING', 'UTF-8'),
        'defaultLocale' => env('APP_DEFAULT_LOCALE', 'en_US'),
        'defaultTimezone' => env('APP_DEFAULT_TIMEZONE', 'UTC'),
        'base' => false,
        'dir' => 'src',
        'webroot' => 'webroot',
        'wwwRoot' => WWW_ROOT,
        //'baseUrl' => env('SCRIPT_NAME'),
        'fullBaseUrl' => false,
        'imageBaseUrl' => 'img/',
        'cssBaseUrl' => 'css/',
        'jsBaseUrl' => 'js/',
        'paths' => [
            'plugins' => [ROOT . DS . 'plugins' . DS],
            'templates' => [ROOT . DS . 'templates' . DS],
            'locales' => [RESOURCES . 'locales' . DS],
        ],
    ],

    /**
     * Security and encryption configuration
     *
     * - salt - A random string used in security hashing methods.
     *   The salt value is also used as the encryption key.
     *   You should treat it as extremely sensitive data.
     */
    'Security' => [
        'salt' => env('SECURITY_SALT', '__SALT__'),
    ],

    /**
     * Apply timestamps with the last modified time to static assets (js, css, images).
     * Will append a querystring parameter containing the time the file was modified.
     * This is useful for busting browser caches.
     *
     * Set to true to apply timestamps when debug is true. Set to 'force' to always
     * enable timestamping regardless of debug value.
     */
    'Asset' => [
        //'timestamp' => true,
        // 'cacheTime' => '+1 year'
    ],

    /**
     * Configure the cache adapters.
     */
    'Cache' => [
        'default' => [
            'className' => 'Cake\Cache\Engine\FileEngine',
            'path' => CACHE,
            'url' => env('CACHE_DEFAULT_URL', null),
        ],

        /**
         * Configure the cache used for general framework caching.
         * Translation cache files are stored with this configuration.
         * Duration will be set to '+2 minutes' in bootstrap.php when debug = true
         * If you set 'className' => 'Null' core cache will be disabled.
         */
        '_cake_core_' => [
            'className' => 'Cake\Cache\Engine\FileEngine',
            'prefix' => 'myapp_cake_core_',
            'path' => CACHE . 'persistent/',
            'serialize' => true,
            'duration' => '+1 years',
            'url' => env('CACHE_CAKECORE_URL', null),
        ],

        /**
         * Configure the cache for model and datasource caches. This cache
         * configuration is used to store schema descriptions, and table listings
         * in connections.
         * Duration will be set to '+2 minutes' in bootstrap.php when debug = true
         */
        '_cake_model_' => [
            'className' => 'Cake\Cache\Engine\FileEngine',
            'prefix' => 'myapp_cake_model_',
            'path' => CACHE . 'models/',
            'serialize' => true,
            'duration' => '+1 years',
            'url' => env('CACHE_CAKEMODEL_URL', null),
        ],

        /**
         * Configure the cache for routes. The cached routes collection is built the
         * first time the routes are processed via `config/routes.php`.
         * Duration will be set to '+2 seconds' in bootstrap.php when debug = true
         */
        '_cake_routes_' => [
            'className' => 'Cake\Cache\Engine\FileEngine',
            'prefix' => 'myapp_cake_routes_',
            'path' => CACHE,
            'serialize' => true,
            'duration' => '+1 years',
            'url' => env('CACHE_CAKEROUTES_URL', null),
        ],
    ],

    /**
     * Configure the Error and Exception handlers used by your application.
     *
     * By default errors are displayed using Debugger, when debug is true and logged
     * by Cake\Log\Log when debug is false.
     *
     * In CLI environments exceptions will be printed to stderr with a backtrace.
     * In web environments an HTML page will be displayed for the exception.
     * With debug true, framework errors like Missing Controller will be displayed.
     * When debug is false, framework errors will be coerced into generic HTTP errors.
     *
     * Options:
     *
     * - `errorLevel` - int - The level of errors you are interested in capturing.
     * - `trace` - boolean - Whether or not backtraces should be included in
     *   logged errors/exceptions.
     * - `log` - boolean - Whether or not you want exceptions logged.
     * - `exceptionRenderer` - string - The class responsible for rendering
     *   uncaught exceptions. If you choose a custom class you should place
     *   the file for that class in src/Error. This class needs to implement a
     *   render method.
     * - `skipLog` - array - List of exceptions to skip for logging. Exceptions that
     *   extend one of the listed exceptions will also be skipped for logging.
     *   E.g.:
     *   `'skipLog' => ['Cake\Network\Exception\NotFoundException', 'Cake\Network\Exception\UnauthorizedException']`
     * - `extraFatalErrorMemory` - int - The number of megabytes to increase
     *   the memory limit by when a fatal error is encountered. This allows
     *   breathing room to complete logging or error handling.
     */
    'Error' => [
        'errorLevel' => E_ALL,
        'exceptionRenderer' => 'Cake\Error\ExceptionRenderer',
        'skipLog' => [],
        'log' => true,
        'trace' => true,
    ],

    /**
     * Email configuration.
     *
     * By defining transports separately from delivery profiles you can easily
     * re-use transport configuration across multiple profiles.
     *
     * You can specify multiple configurations for production, development and
     * testing.
     *
     * Each transport needs a `className`. Valid options are as follows:
     *
     *  Mail   - Send using PHP mail function
     *  Smtp   - Send using SMTP
     *  Debug  - Do not send the email, just return the result
     *
     * You can add custom transports (or override existing transports) by adding the
     * appropriate file to src/Mailer/Transport. Transports should be named
     * 'YourTransport.php', where 'Your' is the name of the transport.
     */
    'EmailTransport' => [
        'default' => [
            'className' => 'Cake\Mailer\Transport\MailTransport',
            /*
             * The following keys are used in SMTP transports:
             */
            'host' => 'localhost',
            'port' => 25,
            'timeout' => 30,
            'username' => null,
            'password' => null,
            'client' => null,
            'tls' => null,
            'url' => env('EMAIL_TRANSPORT_DEFAULT_URL', null),
        ],
    ],

    /**
     * Email delivery profiles
     *
     * Delivery profiles allow you to predefine various properties about email
     * messages from your application and give the settings a name. This saves
     * duplication across your application and makes maintenance and development
     * easier. Each profile accepts a number of keys. See `Cake\Mailer\Email`
     * for more information.
     */
    'Email' => [
        'default' => [
            'transport' => 'default',
            'from' => 'you@localhost',
            //'charset' => 'utf-8',
            //'headerCharset' => 'utf-8',
        ],
    ],

    /**
     * Connection information used by the ORM to connect
     * to your application's datastores.
     *
     * ### Notes
     * - Drivers include Mysql Postgres Sqlite Sqlserver
     *   See vendor\cakephp\cakephp\src\Database\Driver for complete list
     * - Do not use periods in database name - it may lead to error.
     *   See https://github.com/cakephp/cakephp/issues/6471 for details.
     * - 'encoding' is recommended to be set to full UTF-8 4-Byte support.
     *   E.g set it to 'utf8mb4' in MariaDB and MySQL and 'utf8' for any
     *   other RDBMS.
     */
    'Datasources' => [
        'default' => [
            'className' => 'Cake\Database\Connection',
            'driver' => 'Cake\Database\Driver\Mysql',
            'persistent' => false,
            'host' => 'localhost',
            /*
             * CakePHP will use the default DB port based on the driver selected
             * MySQL on MAMP uses port 8889, MAMP users will want to uncomment
             * the following line and set the port accordingly
             */
            //'port' => 'non_standard_port_number',
            'username' => 'my_app',
            'password' => 'secret',
            'database' => 'my_app',
            /*
             * You do not need to set this flag to use full utf-8 encoding (internal default since CakePHP 3.6).
             */
            //'encoding' => 'utf8mb4',
            'timezone' => 'UTC',
            'flags' => [],
            'cacheMetadata' => true,
            'log' => false,

            /**
             * Set identifier quoting to true if you are using reserved words or
             * special characters in your table or column names. Enabling this
             * setting will result in queries built using the Query Builder having
             * identifiers quoted when creating SQL. It should be noted that this
             * decreases performance because each query needs to be traversed and
             * manipulated before being executed.
             */
            'quoteIdentifiers' => false,

            /**
             * During development, if using MySQL < 5.6, uncommenting the
             * following line could boost the speed at which schema metadata is
             * fetched from the database. It can also be set directly with the
             * mysql configuration directive 'innodb_stats_on_metadata = 0'
             * which is the recommended value in production environments
             */
            //'init' => ['SET GLOBAL innodb_stats_on_metadata = 0'],

            'url' => env('DATABASE_URL', null),
        ],

        /**
         * The test connection is used during the test suite.
         */
        'test' => [
            'className' => 'Cake\Database\Connection',
            'driver' => 'Cake\Database\Driver\Mysql',
            'persistent' => false,
            'host' => 'localhost',
            //'port' => 'non_standard_port_number',
            'username' => 'my_app',
            'password' => 'secret',
            'database' => 'test_myapp',
            //'encoding' => 'utf8mb4',
            'timezone' => 'UTC',
            'cacheMetadata' => true,
            'quoteIdentifiers' => false,
            'log' => false,
            //'init' => ['SET GLOBAL innodb_stats_on_metadata = 0'],
            'url' => env('DATABASE_TEST_URL', null),
        ],
    ],

    /**
     * Configures logging options
     */
    'Log' => [
        'debug' => [
            'className' => 'Cake\Log\Engine\FileLog',
            'path' => LOGS,
            'file' => 'debug',
            'url' => env('LOG_DEBUG_URL', null),
            'scopes' => false,
            'levels' => ['notice', 'info', 'debug'],
        ],
        'error' => [
            'className' => 'Cake\Log\Engine\FileLog',
            'path' => LOGS,
            'file' => 'error',
            'url' => env('LOG_ERROR_URL', null),
            'scopes' => false,
            'levels' => ['warning', 'error', 'critical', 'alert', 'emergency'],
        ],
        // To enable this dedicated query log, you need set your datasource's log flag to true
        'queries' => [
            'className' => 'Cake\Log\Engine\FileLog',
            'path' => LOGS,
            'file' => 'queries',
            'url' => env('LOG_QUERIES_URL', null),
            'scopes' => ['queriesLog'],
        ],
    ],

    /**
     * Session configuration.
     *
     * Contains an array of settings to use for session configuration. The
     * `defaults` key is used to define a default preset to use for sessions, any
     * settings declared here will override the settings of the default config.
     *
     * ## Options
     *
     * - `cookie` - The name of the cookie to use. Defaults to 'CAKEPHP'. Avoid using `.` in cookie names,
     *   as PHP will drop sessions from cookies with `.` in the name.
     * - `cookiePath` - The url path for which session cookie is set. Maps to the
     *   `session.cookie_path` php.ini config. Defaults to base path of app.
     * - `timeout` - The time in minutes the session should be valid for.
     *    Pass 0 to disable checking timeout.
     *    Please note that php.ini's session.gc_maxlifetime must be equal to or greater
     *    than the largest Session['timeout'] in all served websites for it to have the
     *    desired effect.
     * - `defaults` - The default configuration set to use as a basis for your session.
     *    There are four built-in options: php, cake, cache, database.
     * - `handler` - Can be used to enable a custom session handler. Expects an
     *    array with at least the `engine` key, being the name of the Session engine
     *    class to use for managing the session. CakePHP bundles the `CacheSession`
     *    and `DatabaseSession` engines.
     * - `ini` - An associative array of additional ini values to set.
     *
     * The built-in `defaults` options are:
     *
     * - 'php' - Uses settings defined in your php.ini.
     * - 'cake' - Saves session files in CakePHP's /tmp directory.
     * - 'database' - Uses CakePHP's database sessions.
     * - 'cache' - Use the Cache class to save sessions.
     *
     * To define a custom session handler, save it at src/Network/Session/<name>.php.
     * Make sure the class implements PHP's `SessionHandlerInterface` and set
     * Session.handler to <name>
     *
     * To use database sessions, load the SQL file located at config/schema/sessions.sql
     */
    'Session' => [
        'defaults' => 'php',
    ],
];
